### PR TITLE
feat: merge-insert supports inserting subset of columns

### LINKED ISSUE #2904: Support inserting in `merge_insert` with a subset of columns
Right now we supporting updating a subset of columns with merge insert. However, we left `when_not_matched_insert_all()` as TODO. (see: https://github.com/lancedb/lance/pull/2639). Users should be able to provide a subset of columns in the merge insert query, and have the missing columns filled in with `null`. (Relatedly, we might also consider something similar with `append`.)

We left this as TODO in large part because we haven't seriously approached null values. When inserting a subset of columns, the missing columns should be filled with nulls. The default V1 format doesn't universally support nulls, so it would not have worked well. Now that V2 is the default, this will make more sense. We should decide again whether we should allow missing (null) fields in fragments, or require that there be all null columns in the data files. We have a vestigial comment in our proto files right now:

https://github.com/lancedb/lance/blob/65b32d4b878de1f47a071e675e05831e8a9eba6c/protos/table.proto#L182-L183

### PR DESCRIPTION
**Pull request by wjones127**
In https://github.com/lancedb/lance/pull/2639 we added support for *updating* subcolumns. In https://github.com/lancedb/lance/pull/3041 we added support for *inserting* subcolumns. This PR adds support for upserting them (or doing insert-if-not-exists).

Closes #2904

## Example

```python
import pyarrow as pa
import lance

table = pa.table({
    "id": range(3),
    "a": [1.0, 2.0, 3.0],
    "c": ["x", "x", "x"]
})
dataset = lance.write_dataset(table, "example")

# Upsert: when_matched_update_all + when_not_matched_insert_all
new_data = pa.table({
    "id": [2, 3],
    "c": ["y", "y"]
})
(
    dataset
    .merge_insert(on="id")
    .when_matched_update_all()
    .when_not_matched_insert_all()
    .execute(new_data)
)
dataset.to_table().to_pandas()
```
```
   id    a  c
0   0  1.0  x
1   1  2.0  x
2   2  3.0  y
3   3  NaN  y
```

```python
# Insert-if-not-exists: when_not_matched_insert_all
new_data = pa.table({
    "id": [3, 4],
    "c": ["z", "z"]
})
(
    dataset
    .merge_insert(on="id")
    .when_not_matched_insert_all()
    .execute(new_data)
)
dataset.to_table().to_pandas()

   id    a  c
0   0  1.0  x
1   1  2.0  x
2   2  3.0  y
3   3  NaN  y
4   4  NaN  z
```

### CODE REVIEW

File: rust/lance/src/dataset/write/merge_insert.rs [hunk 1]
```diff
@@ -787,6 +781,44 @@ impl MergeInsertJob {
...
 
+            async fn handle_new_fragments(
+                dataset: Arc<Dataset>,
+                batches: Vec<RecordBatch>,
+                new_fragments: Arc<Mutex<Vec<Fragment>>>,
+                reservation_size: usize,
+            ) -> Result<usize> {
+                // Batches still have _rowaddr (used elsewhere to merge with existing data)
+                // We need to remove it before writing to Lance files.
+                let num_fields = batches[0].schema().fields().len();
```
Comment (westonpace) [inline rust/lance/src/dataset/write/merge_insert.rs:792]: Not needed for this PR but we should add a drop columns method to schema.

File: python/python/lance/dataset.py [hunk 2]
```diff
@@ -1219,6 +1224,51 @@ def merge_insert(
...
+        {'num_inserted_rows': 1, 'num_updated_rows': 0, 'num_deleted_rows': 0}
+        >>> dataset.to_table().sort_by("a").to_pandas()
+           a  b
+        0  1  a
+        1  2  b
+        2  3  c
+        3  4  z
+
+        You are not required to provide all the columns. If you only want to
+        updated a subset of columns, you can omit columns you don't want to
```
Comment (westonpace) [inline python/python/lance/dataset.py:1250]: `updated` -> `update`

**Review by westonpace — APPROVED** [review]
Nice work!

File: python/python/lance/dataset.py (hunk 2 above)
Comment (wjones127) [reply python/python/lance/dataset.py:1250]:
```suggestion
        update a subset of columns, you can omit columns you don't want to
```

**Pull request merged**